File: lib/allure_report_publisher/lib/providers/info/github.rb
module Publisher
  module Providers
    module Info
      # Github executor info
      #
      class Github < Base
        include Singleton
        include Helpers

        # Executor info
        #
        # @return [Hash]
        def executor(report_url)
          {
            name: "Github",
            type: "github",
            reportName: "AllureReport",
            reportUrl: report_url,
            url: server_url,
            buildUrl: build_url,
            buildOrder: run_id,
            buildName: build_name
          }
        end

        # Github api client
        #
        # @return [Octokit::Client]
        def client
          @client ||= begin
            raise("Missing GITHUB_AUTH_TOKEN environment variable!") unless ENV["GITHUB_AUTH_TOKEN"]

            Octokit::Client.new(access_token: ENV["GITHUB_AUTH_TOKEN"], api_endpoint: ENV["GITHUB_API_URL"])
          end
        end

        # Pull request run
        #
        # @return [Boolean]
        def pr?
          env("GITHUB_EVENT_NAME") == "pull_request"
        end

        # Run id
        #
        # @return [String]
        def run_id
          @run_id ||= env(ALLURE_RUN_ID) || env("GITHUB_RUN_ID")
        end

        # Server url
        #
        # @return [String]
        def server_url
          @server_url ||= env("GITHUB_SERVER_URL")
        end

        # Job name
        #
        # @return [String]
        def build_name
          @build_name ||= env(ALLURE_JOB_NAME) || env("GITHUB_JOB")
        end

        # Github repository
        #
        # @return [String]
        def repository
          @repository ||= env("GITHUB_REPOSITORY")
        end

        # Build url
        #
        # @return [String]
        def build_url
          @build_url ||= "#{server_url}/#{repository}/actions/runs/#{run_id}"
        end
      end
    end
  end
end

File: src/lib/ci/info/github.ts
import {readFileSync} from 'node:fs'

import {GithubCiProvider} from '../providers/github.js'
import {BaseCiInfo} from './base.js'

type GitHubEvent = {
  number: number
  pull_request: {
    head: {
      sha: string
    }
  }
}

export class GithubCiInfo extends BaseCiInfo {
  private _githubEvent?: GitHubEvent

  public executorJson(reportUrl: string): Record<string, string | undefined> {
    return {
      name: 'GitHub',
      type: 'github',
      reportName: 'AllureReport',
      reportUrl,
      url: this.serverUrl,
      buildUrl: this.buildUrl,
      buildOrder: this.runId,
      buildName: this.buildName,
    }
  }

  public get CiProviderClass() {
    return GithubCiProvider
  }

  public get isPR() {
    return process.env.GITHUB_EVENT_NAME === 'pull_request'
  }

  public get runId() {
    return process.env[BaseCiInfo.ALLURE_RUN_ID] || process.env.GITHUB_RUN_ID
  }

  public get serverUrl() {
    return process.env.GITHUB_SERVER_URL
  }

  public get buildName() {
    return (
      process.env[BaseCiInfo.ALLURE_JOB_NAME] ||
      process.env.GITHUB_JOB ||
      (() => {
        throw new Error('Build name not found in environment variables')
      })()
    )
  }

  public get repository() {
    return process.env.GITHUB_REPOSITORY
  }

  public get buildUrl() {
    return `${this.serverUrl}/${this.repository}/actions/runs/${this.runId}`
  }

  public get prId() {
    return this.githubEvent.number
  }

  private get githubEvent(): GitHubEvent {
    if (this._githubEvent) return this._githubEvent

    const eventPath = process.env.GITHUB_EVENT_PATH
    if (!eventPath) {
      throw new Error('Failed to get GitHub event data: GITHUB_EVENT_PATH is not set')
    }

    this._githubEvent = JSON.parse(readFileSync(eventPath, 'utf8')) as GitHubEvent
    return this._githubEvent
  }

  public getPrShaUrl() {
    const sha = this.githubEvent.pull_request?.head?.sha
    if (!sha || !this.prId) return

    const shortSha = sha.slice(0, 8)
    return `[${shortSha}](${this.serverUrl}/${this.repository}/pull/${this.prId}/commits/${sha})`
  }
}
